# MMB8052 Practical 08 - Case Study III: RNA Sequencing Analysis

# Introduction

The analysis of gene expression data, generated by an RNA Sequencing (RNA-Seq) experiment, is one of the most common bioinformatics applications. 

## RNA Sequencing

RNA Sequencing experiments take advantage of a simple assumption – that the number of reads that map to a particular molecule in a next generation experiment is in direct proportion to both its length, and the amount of that molecule that is present in the sample. So if two RNAs of equal length are present at a ratio of 2:1 in two samples, then we will retrieve twice as many reads for one as we will for the other when we sequence the samples. So, given two samples we can normalise for transcript lengths and total number of reads, and work out which species are differentially expressed between the two samples.

RNA-Seq is often seen as a replacement for the very popular microarray approach to studying gene expression, and while the two approaches have different strengths and weaknesses, it is generally true that the data from RNA-Seq correlates better with qPCR data (the “gold standard” for measuring gene expression data) and the sheer depth of the data generated makes RNA-Seq the better choice in many circumstances (assuming the budget allows).

This also means that many of the mistakes, with regards to experimental design, that were made in the early days of microarrays are being repeated in earnest with RNA-Seq experiments. One of the most common false assumptions is that with RNA-Seq you don’t need replicates. This is simply untrue, we still need to model the variance of the population in order to return meaningful statistics and consequently it’s advisable to do at least 4 biological replicates per condition minimum for cell lines, more for mice/rats and many more for humans. Unlike with microarray analysis however technical replicates are seldom required, as despite the stochastic nature of sequencing runs they tend to be highly reproducible.

## RNA-Seq Analysis

After quality assessment (not covered here, but important), RNA-Seq analysis usually comprises 3 (or 4) steps:

1. Align sequencing reads to a reference genome (this step is increasingly optional)
2. Quantify RNA expression against reference gene annotation
3. Analyse RNA counts to define differential expression
4. Functional analysis of differentially expressed genes

Here, we will be focussing on step 3, which is the bulk of the analysis. The final case study in the module will cover functional analysis – which can be applied downstream of many analysis tasks (not just RNA-Seq).

## RNA-Seq Read Alignment



## Quantification of RNA-Seq Data

Modern RNA-Seq analysis procedures tend to avoid the alignment step, which is computationally costly, slow, and results in very large output files. These output files are largely redundant, given that they replicate the contents of the raw FASTQ files.

Rather than alignment, tools which quantify transcript expression via read abundances tend to be favoured in modern analysis. We will be working with data produced by [Salmon](https://combine-lab.github.io/salmon/), a tool which performs _quasi-alignment_ to assign sequencing reads to their transcript of origin. The process by which the quantification estimates we will be working with have been produced will be described below. 

### Quasi-alignment

Instead of the more ‘traditional’ approach of a full read alignment to the reference genome, Salmon uses an efficient, lightweight alignment methodology to assign the reads in an RNA-Seq experiment to their parent transcripts. It then uses these assignments directly to estimate the abundance of each transcript from the reference annotation in the sample.

It is possible for reads to be assigned to more than one transcript – since many transcripts from one gene can share exons. These ambiguous reads are assigned fractionally based on abundance estimates for the possible transcripts made using uniquely assigned reads. For this reason, despite read assignment looking like a discrete, counting problem, we end up with fractional transcript abundances.

### Annotation

There are many sources of gene annotation available. Even for well-annotated organisms, these resources vary in the approaches they take to annotation, and therefore how complete (or otherwise) they appear.

The annotation used in an RNA-Seq experiment is one of the [biggest factors](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-S11-S8) effecting the outcome of the data analysis. The choices available largely depend on the organism that you're working with. For human and mouse experiments, [GENCODE](https://www.gencodegenes.org/) is recommended. The GENCODE gene sets are a combination of the HAVANA manual annotation, and the Ensembl automatic annotation pipeline. The goal is to achieve an accurate and complete annotation of the human (and mouse) genome.

For other organisms, I would recommend working with the Ensembl annotation, where possible. The quality of Ensembl gene sets is consistent, and the data is easily available. The Ensembl FTP server is a convenient resource for this data for many organisms: <http://www.ensembl.org/info/data/ftp/index.html>

For this example experiment, the data was derived from mouse cell lines, so we can download the latest GENCODE mouse gene set, using wget:

```bash 
$ wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M31/gencode.vM31.transcripts.fa.gz
```
